### PR TITLE
Add Automatic-Module-Name to MANIFEST.MF

### LINKED ISSUE #1988: Add Automatic-Module-Name to MANIFEST.MF
This issue is related to but is not a duplicate of https://github.com/cbeust/testng/issues/1808.

One day you will transition Java 9...

Until that day, please add a `Automatic-Module-Name` entry to `MANIFEST.MF` to improve the life of people who have made the move and would like to continue depending on your library. See http://branchandbound.net/blog/java/2017/12/automatic-module-name/ for background information.

Thank you.

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1988

Fixes #1988  .

### Did you remember to?

- [ ] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

**Comment by krmahadevan** [comment]
@cbeust @juherr - I need help with this. This PR correctly adds the entries into the `MANIFEST.MF` file, but when I add `7.0.0-SNAPSHOT` as a dependency in my pom file, none of the TestNG classes are even getting resolved via IntelliJ. 

If this change gets merged then it will cause a lot of confusions amongst our users who also may experience the same problem.
Any idea what should be done to keep intelliJ happy ?

**Comment by krmahadevan** [comment]
Below is how the generated `MANIFEST.MF` looks like

```
10:09 $ unzip -p testng-7.0.0-SNAPSHOT.jar META-INF/MANIFEST.MF
Manifest-Version: 1.0
Bundle-Description: TestNG is a testing framework.
Bundle-License: http://apache.org/licenses/LICENSE-2.0
Automatic-Module-Name: org.testng
Build-Date: 2018-12-26
Bundle-SymbolicName: org.testng
Built-By: krmahadevan
Bundle-ManifestVersion: 2
Bnd-LastModified: 1545798940000
Specification-Title: testng
Import-Package: com.beust.jcommander;version="[1.7.0,3.0.0)";resolutio
 n:=optional,com.google.inject;version="[1.2,1.3)";resolution:=optiona
 l,junit.framework;version="[3.8.1, 5.0.0)";resolution:=optional,org.j
 unit;resolution:=optional,org.junit.runner;resolution:=optional,org.j
 unit.runner.manipulation;resolution:=optional,org.junit.runner.notifi
 cation;resolution:=optional,org.apache.tools.ant;version="[1.7.0, 2.0
 .0)";resolution:=optional,org.apache.tools.ant.taskdefs;version="[1.7
 .0, 2.0.0)";resolution:=optional,org.apache.tools.ant.types;version="
 [1.7.0, 2.0.0)";resolution:=optional,org.apache.tools.ant.types.resou
 rces;version="[1.7.0, 2.0.0)";resolution:=optional,org.apache.tools.a
 nt.types.selectors;version="[1.7.0, 2.0.0)";resolution:=optional,org.
 yaml.snakeyaml;version="[1.6,2.0)";resolution:=optional,org.yaml.snak
 eyaml.constructor;version="[1.6,2.0)";resolution:=optional,org.yaml.s
 nakeyaml.nodes;version="[1.6,2.0)";resolution:=optional,javax.annotat
 ion,javax.script,javax.xml.parsers,org.w3c.dom,org.xml.sax,org.xml.sa
 x.helpers
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-3.4.0.201707252008
Export-Package: org.testng;version="7.0.0.SNAPSHOT";uses:="com.beust.j
 commander,com.google.inject,javax.annotation,org.apache.tools.ant,org
 .apache.tools.ant.taskdefs,org.apache.tools.ant.types,org.testng.anno
 tations,org.testng.internal,org.testng.internal.annotations,org.testn
 g.internal.junit,org.testng.internal.thread.graph,org.testng.xml",org
 .testng.annotations;version="7.0.0.SNAPSHOT";uses:="com.google.inject
 ,org.testng,org.testng.internal.annotations",org.testng.asserts;versi
 on="7.0.0.SNAPSHOT",org.testng.collections;version="7.0.0.SNAPSHOT",o
 rg.testng.internal;version="7.0.0.SNAPSHOT";uses:="com.google.inject,
 javax.annotation,org.testng,org.testng.annotations,org.testng.interna
 l.annotations,org.testng.internal.thread,org.testng.internal.thread.g
 raph,org.testng.junit,org.testng.reporters,org.testng.xml",org.testng
 .internal.annotations;version="7.0.0.SNAPSHOT";uses:="org.testng,org.
 testng.annotations,org.testng.internal,org.testng.xml",org.testng.int
 ernal.collections;version="7.0.0.SNAPSHOT",org.testng.internal.invoke
 rs;version="7.0.0.SNAPSHOT";uses:="org.testng",org.testng.internal.ju
 nit;version="7.0.0.SNAPSHOT",org.testng.internal.reflect;version="7.0
 .0.SNAPSHOT";uses:="org.testng",org.testng.internal.thread;version="7
 .0.0.SNAPSHOT",org.testng.internal.thread.graph;version="7.0.0.SNAPSH
 OT";uses:="org.testng,org.testng.internal",org.testng.junit;version="
 7.0.0.SNAPSHOT";uses:="junit.framework,org.junit.runner,org.testng,or
 g.testng.internal,org.testng.internal.annotations,org.testng.xml",org
 .testng.log;version="7.0.0.SNAPSHOT",org.testng.log4testng;version="7
 .0.0.SNAPSHOT",org.testng.mustache;version="7.0.0.SNAPSHOT",org.testn
 g.reporters;version="7.0.0.SNAPSHOT";uses:="javax.annotation,org.test
 ng,org.testng.internal,org.testng.xml",org.testng.reporters.jq;versio
 n="7.0.0.SNAPSHOT";uses:="org.testng,org.testng.reporters,org.testng.
 xml",org.testng.reporters.util;version="7.0.0.SNAPSHOT";uses:="org.te
 stng",org.testng.util;version="7.0.0.SNAPSHOT";uses:="org.testng",org
 .testng.xml;version="7.0.0.SNAPSHOT";uses:="org.testng,org.testng.rep
 orters,org.testng.xml.dom,org.xml.sax,org.xml.sax.helpers",org.testng
 .xml.dom;version="7.0.0.SNAPSHOT";uses:="javax.xml.parsers,org.testng
 ,org.testng.internal.collections,org.testng.xml,org.w3c.dom,org.xml.s
 ax",org.testng.xml.internal;version="7.0.0.SNAPSHOT";uses:="org.testn
 g.xml"
Bundle-Name: testng
Bundle-Version: 7.0.0.SNAPSHOT
Build-Time: 10:05:38.256+0530
Created-By: 1.8.0_131 (Oracle Corporation 25.131-b11)
Specification-Version: 7.0.0-SNAPSHOT
```

**Comment by cowwoc** [comment]
@krmahadevan Out of curiosity, what happens if you remove `Automatic-Module-Name` from `MANIFEST.MF` and add `7.0.0-SNAPSHOT` as a dependency in my pom file? Does that work as expected?

Also, the project you are adding testng as a dependency for... does it contain `module-info.java` or is it a non-modularized project?

**Comment by krmahadevan** [comment]
> what happens if you remove Automatic-Module-Name from MANIFEST.MF and add
7.0.0-SNAPSHOT as a dependency in my pom file? Does that work as expected?

Without that attribute, I have had no issues. Its only after I added that
attribute to the Manifest file and install the jar locally to the maven
repository and try to consume it in a test project is where I have this as
a problem.

> Also, the project you are adding testng as a dependency for... does it
contain module-info.java or is it a non-modularized project?

All of my work is still on JDK8. So its a non-modularized project.

>

File: build.gradle [hunk 1]
```diff
@@ -120,6 +120,10 @@ jar {
             '!org.testng.*',
             '!com.sun.*',
             '*'
+        instruction "Automatic-Module-Name", "org.testng"
+        attributes(
```
Comment (juherr) [inline build.gradle]: It won't fix your issue but why do we need the duplication (instruction + attributes)?

**Comment by krmahadevan** [comment]
> It won't fix your issue but why do we need the duplication (instruction +
attributes)?

To be honest, I dont know Julien. When I tried adding just the
"instruction", I didnt see anything being added to the MANIFEST.MF and it
still showed some junk values.
The blog link that was shared in the issue recommends adding it up as an
instruction. But it didnt work for me (Perhaps due to a discrepancy with
the Gradle version maybe?)

That was why I went back to adding it as an attribute as well. After I
added it as an attribute, it worked but then cropped up this new problem of
IntelliJ not being able to resolve any of the classes within TestNG, when
consumed via a test project.

>

File: build.gradle (hunk 1 above)
Comment (krmahadevan) [reply build.gradle]: To be honest, I dont know Julien. When I tried adding just the
"instruction", I didnt see anything being added to the MANIFEST.MF and it
still showed some junk values.
The blog link that was shared in the issue recommends adding it up as an
instruction. But it didnt work for me (Perhaps due to a discrepancy with
the Gradle version maybe?)

That was why I went back to adding it as an attribute as well. After I
added it as an attribute, it worked but then cropped up this new problem of
IntelliJ not being able to resolve any of the classes within TestNG, when
consumed via a test project.

**Comment by krmahadevan** [comment]
@juherr 

> Does it work when you run tests on your project with maven/gradle?

It seems to work when i run using maven. I am not very familiar with gradle when compared to maven. So I dont use gradle. The problem seems to be with IntelliJ. I dont use eclipse. So am not sure if the problem exists in eclipse or not.

**Comment by juherr** [comment]
Your issue looks similar to https://youtrack.jetbrains.com/issue/IDEA-183692

I checked the other related issues and I didn't find a better result:
https://youtrack.jetbrains.com/issues?q=tag:%20Jigsaw

Maybe you can help to fix the issue in IntelliJ

File: build.gradle (hunk 1 above)
Comment (juherr) [reply build.gradle]: Please remove the useless line before the merge.

**Review by juherr — APPROVED** [review]

**Comment by krmahadevan** [comment]
@juherr - Merely adding 

`instruction "Automatic-Module-Name", "org.testng"`

doesn't solve the ask of issue #1988 

Here's what the manifest file looks like when i just include `instruction "Automatic-Module-Name", "org.testng"`

```
11:14 $ unzip -p 7.0.0-SNAPSHOT/testng-7.0.0-SNAPSHOT.jar META-INF/MANIFEST.MF
Manifest-Version: 1.0
Bundle-Description: TestNG is a testing framework.
Bundle-License: http://apache.org/licenses/LICENSE-2.0
Automatic-Module-Name: com.acme.mylibrary
Build-Date: 2018-12-28
Bundle-SymbolicName: org.testng
Built-By: krmahadevan
Bundle-ManifestVersion: 2
Bnd-LastModified: 1545975869000
Specification-Title: testng
```

As you can see.. it still adds `Automatic-Module-Name: com.acme.mylibrary`

**Comment by krmahadevan** [comment]
@juherr - Finally managed to get to the bottom of this.

When I add 

```
instruction "Automatic-Module-Name", "org.testng"
```

it doesn't work.

But when I add  (use single quotes instead of double quotes)

```
instruction 'Automatic-Module-Name', 'org.testng'
```

it works. 
The manifest file has the correct entry.

And IntelliJ doesn't complain as well. I guess that adding the module name as an attribute is what irks IntelliJ.

Since the problem is resolved, manifest file shows the correct entry and IntelliJ is happy when I depend on `7.0.0-SNAPSHOT`, I am going ahead and merging this.

**Comment by krmahadevan** [comment]
Here's the complete contents of the manifest file for reference

```
Manifest-Version: 1.0
Bundle-Description: TestNG is a testing framework.
Bundle-License: http://apache.org/licenses/LICENSE-2.0
Automatic-Module-Name: org.testng
Build-Date: 2018-12-28
Bundle-SymbolicName: org.testng
Built-By: krmahadevan
Bundle-ManifestVersion: 2
Bnd-LastModified: 1545979894000
Specification-Title: testng
Import-Package: com.beust.jcommander;version="[1.7.0,3.0.0)";resolutio
 n:=optional,com.google.inject;version="[1.2,1.3)";resolution:=optiona
 l,junit.framework;version="[3.8.1, 5.0.0)";resolution:=optional,org.j
 unit;resolution:=optional,org.junit.runner;resolution:=optional,org.j
 unit.runner.manipulation;resolution:=optional,org.junit.runner.notifi
 cation;resolution:=optional,org.apache.tools.ant;version="[1.7.0, 2.0
 .0)";resolution:=optional,org.apache.tools.ant.taskdefs;version="[1.7
 .0, 2.0.0)";resolution:=optional,org.apache.tools.ant.types;version="
 [1.7.0, 2.0.0)";resolution:=optional,org.apache.tools.ant.types.resou
 rces;version="[1.7.0, 2.0.0)";resolution:=optional,org.apache.tools.a
 nt.types.selectors;version="[1.7.0, 2.0.0)";resolution:=optional,org.
 yaml.snakeyaml;version="[1.6,2.0)";resolution:=optional,org.yaml.snak
 eyaml.constructor;version="[1.6,2.0)";resolution:=optional,org.yaml.s
 nakeyaml.nodes;version="[1.6,2.0)";resolution:=optional,javax.annotat
 ion,javax.script,javax.xml.parsers,org.w3c.dom,org.xml.sax,org.xml.sa
 x.helpers
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-3.4.0.201707252008
Export-Package: org.testng;version="7.0.0.SNAPSHOT";uses:="com.beust.j
 commander,com.google.inject,javax.annotation,org.apache.tools.ant,org
 .apache.tools.ant.taskdefs,org.apache.tools.ant.types,org.testng.anno
 tations,org.testng.internal,org.testng.internal.annotations,org.testn
 g.internal.junit,org.testng.internal.thread.graph,org.testng.xml",org
 .testng.annotations;version="7.0.0.SNAPSHOT";uses:="com.google.inject
 ,org.testng,org.testng.internal.annotations",org.testng.asserts;versi
 on="7.0.0.SNAPSHOT",org.testng.collections;version="7.0.0.SNAPSHOT",o
 rg.testng.internal;version="7.0.0.SNAPSHOT";uses:="com.google.inject,
 javax.annotation,org.testng,org.testng.annotations,org.testng.interna
 l.annotations,org.testng.internal.thread,org.testng.internal.thread.g
 raph,org.testng.junit,org.testng.reporters,org.testng.xml",org.testng
 .internal.annotations;version="7.0.0.SNAPSHOT";uses:="org.testng,org.
 testng.annotations,org.testng.internal,org.testng.xml",org.testng.int
 ernal.collections;version="7.0.0.SNAPSHOT",org.testng.internal.invoke
 rs;version="7.0.0.SNAPSHOT";uses:="org.testng",org.testng.internal.ju
 nit;version="7.0.0.SNAPSHOT",org.testng.internal.reflect;version="7.0
 .0.SNAPSHOT";uses:="org.testng",org.testng.internal.thread;version="7
 .0.0.SNAPSHOT",org.testng.internal.thread.graph;version="7.0.0.SNAPSH
 OT";uses:="org.testng,org.testng.internal",org.testng.junit;version="
 7.0.0.SNAPSHOT";uses:="junit.framework,org.junit.runner,org.testng,or
 g.testng.internal,org.testng.internal.annotations,org.testng.xml",org
 .testng.log;version="7.0.0.SNAPSHOT",org.testng.log4testng;version="7
 .0.0.SNAPSHOT",org.testng.mustache;version="7.0.0.SNAPSHOT",org.testn
 g.reporters;version="7.0.0.SNAPSHOT";uses:="javax.annotation,org.test
 ng,org.testng.internal,org.testng.xml",org.testng.reporters.jq;versio
 n="7.0.0.SNAPSHOT";uses:="org.testng,org.testng.reporters,org.testng.
 xml",org.testng.reporters.util;version="7.0.0.SNAPSHOT";uses:="org.te
 stng",org.testng.util;version="7.0.0.SNAPSHOT";uses:="org.testng",org
 .testng.xml;version="7.0.0.SNAPSHOT";uses:="org.testng,org.testng.rep
 orters,org.testng.xml.dom,org.xml.sax,org.xml.sax.helpers",org.testng
 .xml.dom;version="7.0.0.SNAPSHOT";uses:="javax.xml.parsers,org.testng
 ,org.testng.internal.collections,org.testng.xml,org.w3c.dom,org.xml.s
 ax",org.testng.xml.internal;version="7.0.0.SNAPSHOT";uses:="org.testn
 g.xml"
Bundle-Name: testng
Bundle-Version: 7.0.0.SNAPSHOT
Build-Time: 12:21:30.543+0530
Created-By: 1.8.0_131 (Oracle Corporation 25.131-b11)
Specification-Version: 7.0.0-SNAPSHOT

```

**Comment by cowwoc** [comment]
> @juherr - Finally managed to get to the bottom of this.

Gotta love these kind of issues... Thanks for all your hard work!

**Pull request merged**